### ISSUE TITLE
Misleading error message when mistyped pattern matching in closure's argument

### ISSUE THREAD

**Issue by yeputons** [rust-lang/rust#96977]
Given the following code: ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aee5344af3e9eb717a5b02db6801707d))

```rust
fn main() {
    let _ =
        std::collections::HashMap::<String, i32>::new()
            .iter()
            .map(|(k, v | format!("{}->{}", k, v));
}
```

The current output is:

```
error: mismatched closing delimiter: `}`
 --> src/main.rs:5:17
  |
1 | fn main() {
  |           - closing delimiter possibly meant for this
...
5 |             .map(|(k, v | format!("{}->{}", k, v));
  |                 ^ unclosed delimiter
6 | }
  | ^ mismatched closing delimiter

error: expected one of `,` or `:`, found `;`
 --> src/main.rs:5:51
  |
5 |             .map(|(k, v | format!("{}->{}", k, v));
  |                                                   ^ expected one of `,` or `:`
```

Ideally the output should look like:

```
error: expected one of `)`, `,`, `@`, or `|`, found `=>`
 --> src/main.rs:5:24
5 |             .map(|(k, v | format!("{}->{}", k, v));
  |                   ^    -^
  |                   |    |
  |                   |    help: `)` may belong here
  |                   ^ unclosed delimiter
```

Almost like it currently does for a similar `match` mistype: ([playgroud](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f2beb488739f577a0cf62c6915e9eca5))

```rust
fn main() {
    match (1, 2) {
        (k, v => 0
    }
}
```

Although I'd prefer if it is a single error message, not two, like currently:

```
error: expected one of `)`, `,`, `@`, or `|`, found `=>`
 --> src/main.rs:3:9
  |
3 |         (k, v => 0
  |         ^    -^^
  |         |    |
  |         |    help: `)` may belong here
  |         unclosed delimiter

error: expected one of `,`, `.`, `?`, `}`, or an operator, found `)`
 --> src/main.rs:4:5
  |
3 |         (k, v => 0
  |               --  - expected one of `,`, `.`, `?`, `}`, or an operator
  |               |
  |               while parsing the `match` arm starting here
4 |     }
  |     ^ unexpected token
```

**Comment by yeputons** [comment]
A bit of insight in why that may be hard:

`map(  |(k, v)| some_macro!(....)  )` has a closure which takes a single argument: a pair of `k` and `v`

`map(  |(k, v | some_macro!(...))  ` is a prefix of a closure, specifically:

`map(  |(k, v | some_macro!(...)), other| k + v + other   )`. Here the closure takes two arguments. The second one is `other`, the first one is a pair of `k` and `v | some_macro!(...)`, the latter being a pattern.

So it's basically hard to tell whether `|` should be an end of a closure syntax or an alternative operator. We can't even count the balance of `|`s, as there may be an arbitrary (even/odd) number of them in a correct code.
